pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+            Ok(().into())
+        }
+    }
+}
+
+impl<T: Config> LiquidStakingProtocol<T::AccountId> for Pallet<T> {
+    // After confirmed bond on relaychain,
+    // after update exchangerate (record_reward),
+    // and then mint/deposit xKSM.
+    fn stake(who: &T::AccountId, amount: Balance) -> DispatchResultWithPostInfo {
+        //todo reserve, insurance pool
+        T::Currency::transfer(T::StakingCurrency::get(), who, &Self::account_id(), amount)?;
+
+        MatchingQueue::<T>::try_mutate(
+            who,
+            &Self::current_era(),
+            |user_buffer| -> DispatchResult {
+                user_buffer.total_stake_amount = user_buffer
+                    .total_stake_amount
+                    .checked_add(amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPool::<T>::try_mutate(&Self::current_era(), |pool_buffer| -> DispatchResult {
+            pool_buffer.total_stake_amount = pool_buffer
+                .total_stake_amount
+                .checked_add(amount)
+                .ok_or(ArithmeticError::Overflow)?;
+
+            Ok(())
+        })?;
+        Self::deposit_event(Event::Staked(who.clone(), amount));
+        Ok(().into())
+    }
+
+    // After confirmed unbond on relaychain,
+    // and then burn/withdraw xKSM.
+    // before update exchangerate (record_reward)
+    fn unstake(who: &T::AccountId, amount: Balance) -> DispatchResultWithPostInfo {
+        // can not burn directly because we have match mechanism
+        T::Currency::transfer(T::LiquidCurrency::get(), who, &Self::account_id(), amount)?;
+
+        let exchange_rate = ExchangeRate::<T>::get();
+        let asset_amount = exchange_rate
+            .checked_mul_int(amount)
+            .ok_or(Error::<T>::InvalidExchangeRate)?;
+
+        MatchingQueue::<T>::try_mutate(
+            who,
+            &Self::current_era(),
+            |user_buffer| -> DispatchResult {
+                user_buffer.total_unstake_amount = user_buffer
+                    .total_unstake_amount
+                    .checked_add(asset_amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPool::<T>::try_mutate(&Self::current_era(), |pool_buffer| -> DispatchResult {
+            pool_buffer.total_unstake_amount = pool_buffer
+                .total_unstake_amount
+                .checked_add(asset_amount)
+                .ok_or(ArithmeticError::Overflow)?;
+
+            Ok(())
+        })?;
+        Self::deposit_event(Event::Unstaked(who.clone(), amount, asset_amount));
+        Ok(().into())
+    }
+
+    fn claim(who: &T::AccountId) -> DispatchResultWithPostInfo {
+        let mut withdrawable_stake_amount = 0u128;
+        let mut withdrawable_unstake_amount = 0u128;
+        let mut remove = vec![];
+        MatchingQueue::<T>::iter_prefix(who).for_each(|(era_index, user_buffer)| {
```

Comment:
It's such a long function, and there are too many if-else nesting. Replace code in the matching branch with a function and extract common check items.